[Feature/Fix]: Interface injections, and minor bugfixes

marchermans opened the pull request:
### TLDR:
- Implements interface injection
- Adds tests for interface injection
- **Removes** the ability for the gradle script to define ATs, as it never really worked and was a headache

---

### Changes
- Implement interface injection:
  - Implement a way to add additional source files
  - Remove the additional source file outputs from the compile result
  - Handle internal processing on the pipeline scope of additional processing files.
- Implement preprocessing of the recompile step
- Change the problem reporter to be logger based so that Gradle 8.10 does not fall on its face, there is now more then one kind of ProblemReporter implementation in the service stack, so it has no idea what to inject and I can't be bothered to figure that one out, so a simple logger with the same output will do for now.
  - This behaviour can be changed using a convention, and the legacy integrated reporter is still available if desired.

marchermans committed:
Introduce an isolated problem reporter to provide compatibility with gradle.

Works around: Gradle/Gradle#30203

marchermans committed:
Bricked GDI But working setup

marchermans committed:
Merge branch 'refs/heads/NG_7.0' into feature/interface-injection

marchermans committed:
Initial working version

marchermans committed:
Several fixes for broken tests

marchermans committed:
Fix vanilla user transformer not detecting missing files and not running.

lukebemish commented on common/src/main/java/net/neoforged/gradle/common/interfaceinjection/InterfaceInjectionPublishing.java:
```diff
@@ -0,0 +1,56 @@
+    public static final String INTERFACE_INJECTION_ELEMENTS_CONFIGURATION = "InterfaceInjectionElements";
+    public static final String INTERFACE_INJECTION_API_CONFIGURATION = "InterfaceInjectionApi";
+    public static final String INTERFACE_INJECTION_CONFIGURATION = "InterfaceInjection";
+    public static final String INTERFACE_INJECTION_CATEGORY = "InterfaceInjection";
```
This should probably be lowercase, to match both MDG, which uses `interfaceinjection`, and gradle itself, for which all category attribute values are lowercase.

marchermans committed:
Fix a copy paste error caused by replacement

lukebemish commented:
Nope, still not it -- should be fully lowercase, `interfaceinjection`.

marchermans committed:
Lowercase completely

marchermans committed:
Swap back to detached configurations for internal configurations.
This prevents a CME caused by the resolution of the dependent projects configurations triggering the creation of new configuration in its configuration container.
Creating detached configurations, and using them either as tool resolvers, or as carriers for dependency objects which are then attached to an outer configuration should prevent this.
Added tests that cover the new corver cases, and validate publishing logic

marchermans committed:
Fix first integration test that failed because of a difference with how Ingredients are constructed.

marchermans committed:
Fix config cache recompile issues.

marchermans committed:
Update README.md Merging since all tests pass.
